# PHLASH

## Introduction
*Phlash* is an interactive web application that accelerates the manual process of bacteriophage genome annotation. To use the application, click [here](https://bioapps.byu.edu/phlash). Please watch this [demo](https://www.loom.com/share/90814d167d5148378862f4b92abfbfb5) to get a preview of how it works!

## Background
Bacteriophages are diverse, ubiquitous organisms that create the microbial driven world we currently live in. Hundreds of students and faculty members at universities across the nation participate in the [SEA-PHAGES](https://seaphages.org/) program to discover and analyze uncharted bacteria, causing phage genomes to be added to GenBank at an exponential rate. 

In order to annotate genomes, researchers currently use bioinformatic annotation tools—[GeneMark](https://www.ncbi.nlm.nih.gov/pubmed/15980510) and [DNA Master](https://phagesdb.org/DNAMaster/)—to identify at least 80% of genes correctly. While these tools are a vast improvement from past approaches to genome annotation, each software still comes with its own limitations, including access and usability. Even after auto-annotation, researchers are required to manually reexamine the phage’s entire genome and add, modify, or delete genes depending on the data provided from the aforementioned tools. Such a task is not always easy and predicated on varying principles, including manual inspection between possible start sites, relationship to the closest upstream gene, highest coding potential, longer possible frame, etc. Although bacteriophage genomes are relatively small compared to eukaryotic genomes, this process is manually intensive and can consume hours of research that can otherwise be used more productively. 

This tedious nature of genome annotation is a bottleneck to analysis of phage genomes. Once researchers complete genome annotation, they perform, for example, [comparative analysis](https://www.ncbi.nlm.nih.gov/books/NBK20253/) of phage genomes, which reveals common architectural themes in genomes, unique protein functions, and genetic diversity. Such information furthers existing biological knowledge about the phage’s phamilies, as well as its relationships to other organisms, namely humans. The results of current phage genome annotations have extensively contributed to a variety of [phage-based therapies](https://en.wikipedia.org/wiki/Phage_therapy) in medicine and agriculture, especially regarding drug-resistant and long-term bacterial infections. Additionally, the human body’s abundance of phages and other viruses is recognized as an important mediator of human health and well-being, therefore emphasizing the need to continue increasing knowledge about phage communities. 

We developed *Phlash* to allow phage genome annotation to be as quick and efficient as possible. *Phlash* reduces the manual challenges by combining functionalities of [GeneMarkS](https://academic.oup.com/nar/article/29/12/2607/1034721?login=true), [Glimmer3](http://ccb.jhu.edu/papers/glimmer3.pdf), [Aragorn](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC373265/), and [PHANOTATE](https://academic.oup.com/bioinformatics/article/35/22/4537/5480131) and outputting submission-ready GenBank files. By dramatically accelerating the pace of bacteriophage genome annotation, we hope that this application will enable phage hunters will then be able to spend less time annotating and more time contributing to our knowledge on the genetic diversity of bacteriophages and their implications. 

## Project Setup
The best way to run this web application is via [Docker](https://www.docker.com/). For more information on how to set up the client and the server see the README files in the [front-end](https://github.com/srp33/Phlash/blob/master/front-end/README.md) and [back-end](https://github.com/srp33/Phlash/blob/master/back-end/README.md) directories.
